### PR TITLE
Support multiple identity providers

### PR DESCRIPTION
**Pull request by mathildemouw**
This feature will allow the devise_saml_authenticatable gem to [support multiple identity providers (IdPs)](https://github.com/apokalipto/devise_saml_authenticatable/issues/44). 

The `DeviseSamlAuthenticatable::SamlConfig` module currently looks for IdP settings in either Devise.saml_configure in config/initializers/devise.rb or in config/idp.yml. This PR adds a third way for `SamlConfig` to find IdP settings: in an adapter class that takes an IdP entity id and returns a hash of settings for that specific IdP. The IdP-specific settings are then merged into the existing saml_config settings so we do not lose any default settings. This look up method will be used when the `config.idp_settings_adapter` is set in config/initializers/devise.rb.

In the `Devise::SamlSessionsController` actions, we now parse SAML assertions for the IdP entity id to pass to an adapter. The `Strategies::SamlAuthenticatable#authenticate!` method now looks up the specific IdP when an IdP entity id is provided.

Our solution leaves it up to the user to determine how to implement the adapter class, allowing for any mechanism to be used to fetch the idp settings from wherever they are stored. This was important for our use case and is not currently supported in the suggested workarounds in other SAML SP gems, as far as we could find.

### CODE REVIEW

File: lib/devise_saml_authenticatable/saml_config.rb [hunk 1]
```diff
@@ -1,16 +1,39 @@
...
       idp_config_path = "#{Rails.root}/config/idp.yml"
-      # Support 0.0.x-style configuration via a YAML file
+
       if File.exists?(idp_config_path)
-        Devise.saml_config = OneLogin::RubySaml::Settings.new(YAML.load(File.read(idp_config_path))[Rails.env])
+        @file_based_config ||= OneLogin::RubySaml::Settings.new(YAML.load(File.read(idp_config_path))[Rails.env])
+      end
+    end
+
+    def adapter_based_config(idp_entity_id)
```
Comment (mathildemouw) [inline lib/devise_saml_authenticatable/saml_config.rb]: This method merges the adapter config settings into the existing `saml_config` so we don't lose any default settings.

File: app/controllers/devise/saml_sessions_controller.rb [hunk 2]
```diff
@@ -44,8 +46,11 @@ def after_sign_out_path_for(_)
     request.create(saml_config)
   end
 
-  def generate_idp_logout_response(logout_request)
-    logout_request_id = logout_request.id
+  def generate_idp_logout_response(saml_config, logout_request_id)
     OneLogin::RubySaml::SloLogoutresponse.new.create(saml_config, logout_request_id, nil)
   end
+
+  def get_logout_idp_entity_id(saml_request)
```
Comment (mathildemouw) [inline app/controllers/devise/saml_sessions_controller.rb]: This method parses a SAML logout request for the IdP entity id.

File: lib/devise_saml_authenticatable/saml_config.rb [hunk 3]
```diff
@@ -1,16 +1,39 @@
...
+
+        if config.respond_to? acc
+          config.send(acc, v)
+        end
       end
+      config
+    end
 
-      @saml_config = Devise.saml_config
+    def get_idp_entity_id(saml_response)
```
Comment (mathildemouw) [inline lib/devise_saml_authenticatable/saml_config.rb]: This method parses a SAMLResponse for the IdP entity id.

File: lib/devise_saml_authenticatable/strategy.rb [hunk 4]
```diff
@@ -6,15 +6,14 @@ class SamlAuthenticatable < Authenticatable
       include DeviseSamlAuthenticatable::SamlConfig
       def valid?
         if params[:SAMLResponse]
-          response = OneLogin::RubySaml::Logoutresponse.new(params[:SAMLResponse], saml_config)
-          !(response.response.include? 'LogoutResponse')
+          OneLogin::RubySaml::Response.new(params[:SAMLResponse])
```
Comment (mathildemouw) [inline lib/devise_saml_authenticatable/strategy.rb]: It turned out that warden only uses the `authenticate!` and optional `valid?` method in the context of authentication (https://github.com/hassox/warden/wiki/Strategies), so we were able to remove the logout-specific behavior here.

File: spec/controllers/devise/saml_sessions_controller_spec.rb [hunk 5]
```diff
@@ -17,25 +17,79 @@ def require_no_authentication
...
+    let(:saml_response) { File.read(File.join(File.dirname(__FILE__), '../../support', 'response_encrypted_nameid.xml.base64')) }
+
+    context "when using the default saml config" do
+      it "redirects to the IdP SSO target url" do
+        get :new, "SAMLResponse" => saml_response
+        expect(response).to redirect_to(%r(\Ahttp://localhost:8009/saml/auth\?SAMLRequest=))
+      end
+    end
+
+    context "with a specified idp" do
```
Comment (mathildemouw) [inline spec/controllers/devise/saml_sessions_controller_spec.rb]: In this new context, `new` redirects to the SSO target url of the IdP that corresponds to the `Issuer` in the parsed SAML when `Devise.idp_settings_adapter` is set.

File: spec/controllers/devise/saml_sessions_controller_spec.rb [hunk 6]
```diff
@@ -17,25 +17,79 @@ def require_no_authentication
...
+        get :new, "SAMLResponse" => saml_response
+        expect(response).to redirect_to(%r(\Ahttp://idp_sso_url\?SAMLRequest=))
+      end
     end
   end
 
   describe '#metadata' do
-    it 'generates metadata' do
-      get :metadata
+    context "with the default configuration" do
```
Comment (mathildemouw) [inline spec/controllers/devise/saml_sessions_controller_spec.rb]: The metadata returned from this endpoint only needs SP-related information, so we added a new testing context but didn't have to change the controller action to support IdP-specific behavior.

File: spec/controllers/devise/saml_sessions_controller_spec.rb [hunk 7]
```diff
@@ -75,6 +129,20 @@ def require_no_authentication
       expect(response).to redirect_to '/users/saml/sign_in'
     end
 
+    context "with a specified idp" do
```
Comment (mathildemouw) [inline spec/controllers/devise/saml_sessions_controller_spec.rb]: In this new context, `idp_sign_ou` redirects to the SSO target url of the IdP that corresponds to the `Issuer` in the parsed SAML logout request when `Devise.idp_settings_adapter` is set.

File: spec/devise_saml_authenticatable/strategy_spec.rb [hunk 8]
```diff
@@ -35,15 +31,55 @@
...
+      before do
+        Devise.idp_settings_adapter = idp_providers_adapter
+        allow(idp_providers_adapter).to receive(:settings).and_return({})
+      end
+
+      it "is valid" do
+        expect(strategy).to be_valid
+      end
+
+      it "authenticates with the response for the corresponding idp" do
```
Comment (mathildemouw) [inline spec/devise_saml_authenticatable/strategy_spec.rb]: The authenticate method new authenticates for the corresponding IdP after the adapter looks up the settings by entity id. This method is called in the controller `create` action. The create action is not tested in the `SamlSessionsController` specs, however, because it inherits all of its behavior from `Devise::SessionsController`

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -15,6 +15,19 @@
     # `true` in RSpec 4.
     mocks.verify_partial_doubles = true
   end
+
+  config.before :each do
```
Comment (mathildemouw) [inline spec/spec_helper.rb]: Many of the specs change the global Devise.setup config settings so we pulled out before and after reset behavior into the spec helper.

File: spec/support/response_encrypted_nameid.xml.base64 [hunk 10]
```diff
@@ -0,0 +1 @@
...YzpDaXBoZXJEYXRhPg0KICAgICAgPHhlbmM6Q2lwaGVyVmFsdWU+dFdQZEV1dXZmSjh3WVBhOFVUQTRvR2htRENQTjFhQzVkUUFEN0g5SkhWQm5VS3Y0UkljNEQ3SnVJem12bXlyalZGWmRGNW15K3cvUGd3dWlOVGdpOUxid01iSW5adW1HbDhlSndFblBaVXBPQ0w1dDNXbEdKbU85OVVNejZQUVNLeGlGSU1DYzcrQXlRQmpjdTEzaUxWeU5TbFQyWDMxRXBOaW5jQ3FzSldvPTwveGVuYzpDaXBoZXJWYWx1ZT4NCiAgIDwveGVuYzpDaXBoZXJEYXRhPg0KPC94ZW5jOkVuY3J5cHRlZERhdGE+PC9zYW1sOkVuY3J5cHRlZElEPjwvc2FtbDpTdWJqZWN0PjxzYW1sOkNvbmRpdGlvbnMgTm90QmVmb3JlPSIyMDExLTA2LTA0VDAyOjE3OjAyWiIgTm90T25PckFmdGVyPSIyMDMwLTA2LTA0VDAyOjI3OjAyWiI+PHNhbWw6QXVkaWVuY2VSZXN0cmljdGlvbj48c2FtbDpBdWRpZW5jZT5odHRwczovL3NvbWVvbmUuZXhhbXBsZS5jb20vYXVkaWVuY2U8L3NhbWw6QXVkaWVuY2U+PC9zYW1sOkF1ZGllbmNlUmVzdHJpY3Rpb24+PC9zYW1sOkNvbmRpdGlvbnM+PHNhbWw6QXV0aG5TdGF0ZW1lbnQgQXV0aG5JbnN0YW50PSIyMDE0LTA2LTA0VDAyOjIyOjAyWiIgU2Vzc2lvbk5vdE9uT3JBZnRlcj0iMjAzMC0wNi0wNVQwMjoyMjowMloiIFNlc3Npb25JbmRleD0iXzE2ZjU3MGZiYzAzMTUwMDdhMDM1NWRmZWE2YjNjNDZjIj48c2FtbDpBdXRobkNvbnRleHQ+PHNhbWw6QXV0aG5Db250ZXh0Q2xhc3NSZWY+dXJuOm9hc2lzOm5hbWVzOnRjOlNBTUw6Mi4wOmFjOmNsYXNzZXM6UGFzc3dvcmRQcm90ZWN0ZWRUcmFuc3BvcnQ8L3NhbWw6QXV0aG5Db250ZXh0Q2xhc3NSZWY+PC9zYW1sOkF1dGhuQ29udGV4dD48L3NhbWw6QXV0aG5TdGF0ZW1lbnQ+PC9zYW1sOkFzc2VydGlvbj48L3NhbWxwOlJlc3BvbnNlPg==
```
Comment (mathildemouw) [inline spec/support/response_encrypted_nameid.xml.base64]: This test responses comes from the [ruby-saml gem](https://github.com/onelogin/ruby-saml/blob/master/test/responses/response_encrypted_nameid.xml.base64).

**Comment by mathildemouw** [comment]
I added some comments that I hope will help with code review!

**Comment by adamstegman** [comment]
I'm still reviewing the code, but this seems worth a new integration test. Could you add one?

It's not terribly straightforward but basically it generates an SP and IdP application for each test, so the configuration is in Rails templates and controlled by environment variables during app creation. For this case we might need to add a second IdP application (we could probably use the same template and add configuration for the issuer and maybe particular attributes to distinguish the two).

I'm happy to help if you'd like.

**Comment by mathildemouw** [comment]
@adamstegman Thanks for taking a look at the PR! Let me know if you have any questions. I can take a crack at adding an integration spec for this feature. I'll reach out if I have any questions about the test setup.

**Comment by mathildemouw** [comment]
@adamstegman For the integration specs, I'm having trouble getting the string interpolation of environment variables pattern in the sp_template.rb to work for `config.idp_settings_adapter` because this key is supposed to point to an object. I get an `uninitialized constant IdpSettingsAdapter (NameError)` message for the class I'm trying to set it to. Any idea where I should be defining the adapter class for the test SP app?

**Comment by adamstegman** [comment]
It would need to be in the generated application itself, so probably a file like the `saml_idp_controller.rb.erb` template, that is placed into the generated app:

```
template File.expand_path('../idp_settings_adapter.rb.erb', __FILE__), 'app/models/idp_settings_adapter.rb'
```

File: spec/devise_saml_authenticatable/saml_config_spec.rb [hunk 11]
```diff
@@ -1,17 +1,9 @@
 require 'spec_helper'
 
 describe DeviseSamlAuthenticatable::SamlConfig do
-  subject(:saml_config) { controller.saml_config }
+  let(:saml_config) { controller.saml_config }
```
Comment (adamstegman) [inline spec/devise_saml_authenticatable/saml_config_spec.rb]: Why this change?

**Comment by adamstegman** [comment]
The code looks good.

I'm a little curious about the API. Why use an object instead of configuring the hash statically? Is it expected that the IdP configuration would adjust at runtime?

**Comment by mathildemouw** [comment]
That's correct. In our case, we want to store our IdP configurations in the database, so the adapter interface needed to allow for the IdP configuration to come from anywhere and be looked up on the fly.

File: spec/devise_saml_authenticatable/saml_config_spec.rb (hunk 11 above)
Comment (mathildemouw) [inline spec/devise_saml_authenticatable/saml_config_spec.rb]: We weren't sure if it would be cleaner to redefine `subject` on line 26, or just let-ize `saml_config` for the whole suite. It felt strange to keep subject when we end up changing it further on in the spec.

**Comment by adamstegman** [comment]
@mathildemouw if you'd like I can definitely help with the integration test.

**Comment by mathildemouw** [comment]
@adamstegman that would be awesome. Do you have screen hero so we could pair on it?

**Comment by adamstegman** [comment]
@mathildemouw I do have it. Can you email me at me@adamstegman.com so we can coordinate a time? I'm in the US Pacific time zone.

**Comment by mathildemouw** [comment]
Integration test added!

**Comment by mathildemouw** [comment]
@adamstegman looks like all the checks passed!
